### PR TITLE
Implement osu target mod

### LINKED ISSUE #8549: Implement Target mod

### PR DESCRIPTION
**Pull request by hlysine**
Closes #8549

Depends on:
- [x] #13626
- [x] #13645

**To do:**
- [x] Add a seed for random circle generation (similar to #12568)
- [x] ~Switch to results screen after failing~ Show the fail overlay as usual instead
- [x] Replace sample for metronome
- [x] Match jump distance and circle scaling to those in cutting-edge
- [x] Avoid overlapping circles

**Differences with cutting-edge:**
- Cutting-edge uses the first hit object in the map as the first beat, which feels strange if the map starts on an offbeat. This implementation uses the start time of the first timing point (or the one that is active at the first hit object) instead. Example: the first few seconds of https://osu.ppy.sh/beatmapsets/1146827#osu/2394642
- Cutting-edge does not generate circles during breaks. This implementation does not generate circles from the start of a break to the start of the first hit object after the break. This makes the mod work well with compilation maps.
- Cutting-edge groups every 8 circles into a combo. This implementation respects the combos in the original map.

### CODE REVIEW

**Comment by smoogipoo** [comment]
I'm not sure that this should be implemented as a mod (as opposed to an entire ruleset), with it being so different to the point that even the sprites used should be different.

**Comment by peppy** [comment]
I think at most the sprites would be replaced via a skin implementation (maybe restoring what stable did when we have more advanced skinning in place).

Is fine to go ahead without any visual replacements for now, for sure.

**Comment by Henry-YSLin** [comment]
The seed setting is just tentative. I'm guessing it won't work for multi if the user doesn't set an explicit seed. My idea to fix this is to add a new IApplicable_ type that allows the mod to read `MultiplayerRoomSettings` from `MultiplayerClient` and use `PlaylistItemId` as the default seed. Does that sound alright?

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (PercyDan54) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:61]: Why not `IApplicableFailOverride`

**Comment by peppy** [comment]
Have you checked how other mods handle seeing (see `ModRandom`, which recently got this added at an abstract level)? You can likely just derive `ModRandom` for your implementation, or if that isn't possible for some reason, abstract the seed logic out of `ModRandom` into a new `ModWithSeed` base class.

**Comment by Henry-YSLin** [comment]
That was exactly what I was thinking, but `OsuModTarget` is already inheriting from `ModWithVisibilityAdjustment`, so the closest I can do is an `IHasSeed` interface.

**Comment by peppy** [comment]
If it helps getting things accurate, here's the source for the original target practice mod. Note that it isn't necessary to match it 1:1 as it's only present on cutting-edge and not really considered an in-stone implementation.

https://gist.github.com/peppy/ac4b905861d674e17e3e1a88aa7cc05e

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:142]: constants like this need to be extracted out to `private const` fields, with xmldoc where the name isn't already explaining what they mean.

**Comment by Henry-YSLin** [comment]
Is there a sample for the metronome in this mod? I can't find one in osu-resources.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (peppy) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]:
```suggestion
        /// The maximum allowed jump distance after multipliers are applied
```

Incorrect spacing, also applies to all other xmldoc.

Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: Already fixed. I had the wrong coding style setting in Rider.

**Comment by Henry-YSLin** [comment]
The only unimplemented mechanic now is the shattering of target sprites when they are hit, which should come together with the visual replacements later. Aside from that, this mod should be ready for a full review.

**Comment by bdach** [comment]
Just by scrolling this diff I can see that this pull is doing several things at once, namely:

* implementing the target mod itself
* adding a new mod API in `IApplicableFailOverride.DisplayResultsOnFail` (and seemingly setting it to `false` for anything that isn't the target mod)
* adding a new `IHasSeed` interface and a control for it

And partly due to the above, has grown to XL size without necessarily needing to and has several unresolved merge conflicts. Unsure how much of a fuss to make about this but splitting this into chunks would definitely make this more digestible given that the actual mod implementation itself is almost 500 lines on its own.

**Comment by Henry-YSLin** [comment]
Noted. I will open new PRs for `IApplicableFailOverride.DisplayResultsOnFail` and `IHasSeed`. After they are merged, I will come back to target mod and hopefully make the diff less intimidating.

**Comment by bdach** [comment]
As far as I can see the rejected changes from #13616 were not originally a part of this pull. Were they required in some capacity? They don't seem to be.

**Comment by Henry-YSLin** [comment]
No, they weren't required. I just thought the naming was confusing while I was implementing support for `IApplicableFailOverride.DisplayResultsOnFail` in the newly rewritten `Player`.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:196]: Should be `endObj.GetEndTime()` (it's an extension method). Also we don't use switch expressions in this repository.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:211]: * This should really be a method rather than an inline linq lambda. And one with preferably longer variable names.
* This method of generating beats is susceptible to drifting out of sync due to accumulating floating point error. To avoid that this should be done via multiplication, e.g. `startTime + i * beatLength` where `i` increments up until the entire expression exceeds `endTime`.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:217]: This needs to be split out more, I can't follow what this is attempting to do at all. Especially not with variable names of the `b`, `y`, `x` kind.

Also all of these `Precision` calls around are using the default precision of `1e-3` for floats which seems too little for timing objects. It should probably be `1` (millisecond).

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:407]: The modulo op on floats is dangerous, especially if we're talking about precision. With fractions that are fundamentally unrepresentable using floating point arithmetic you can get unexpected results. Example:

```
12.1 % 0.1 = 0.09999999999999898
 0.2 % 0.1 = 0
-0.1 % 0.1 = -0
-0.3 % 0.1 = -0.09999999999999998
```

So doing things this way may not prove to be a good idea. You may want to consider doing division here and checking the fractional part of the dividend or something of the sort. Compare mania's `DistanceObjectPatternGenerator.nodeSamplesAt()`.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:401]: Should really be using `IHasPathWithRepeats` for this sort of thing.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:246]: What's the decrement for? And the `lastSampleIdx >= Count` check, too?

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:268]: These linq expressions direly need actual explanations for what they're doing. I think this is trying to fetch the closest object preceding `x`? I can't get to that from "follow the combo indices in the original beatmap".

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:271]: If this is going to be done, wouldn't it be simpler to get a list of all start times of objects that have `NewCombo = true`, and then just iteratively reindex hitobjects up to the next new combo?

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:292]: I'm pretty sure this isn't doing what is intended. If `Seed.Value` is null, then calling `.GetValueOrDefault()` returns 0. So this is not in fact using some random seed, it's using exactly seed 0 every time.

`new Random(0)` is not equivalent to `new Random()`. The latter [uses tick count for initialisation](https://referencesource.microsoft.com/#mscorlib/system/random.cs,52).

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:296]: Please don't reference osuTK constants for 2*pi. If it is really intended to have it with single precision, you can use a (method-local) const.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:324]: Use `MathF` methods rather than `(float)Math`

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:425]: Why is this operating on the entire list and iindirectly iterating over it via linq rather than taking a `Span` of the few objects that are actually needed here or something of the sort?

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:432]: These are lifted verbatim from `OsuModRandom` and probably should be split to a common helper.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
The above review was mostly surface-level, without digging too much into the code. Readability improvements are direly needed; I think that linq has been (over)used in this code to the point of obfuscating logic.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: The decrement was originally needed because the `while` always loops past the target hit object. I've reworked the logic a bit so that it is no longer needed.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: I have tried this approach, but the code is much more complex because there are cases where an entire combo in the original map gets skipped over. See the newly added comments for details.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:287]: `Seed.Value` can never be null at this point because it has already been replaced by a random value if the user doesn't specify a seed. I should move this to a better place to avoid the confusion.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: I've refactored the method, but I'm not sure if you can take a `Span` from a `List`, so the method is accepting an `IEnumerable` for now.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: There should be an assertion that the seed is not null in that case, instead of falling back to a default that is never supposed to actually happen

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:488]: This was originally on the to-do list but got forgotten somehow. I will open a new PR for this.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:323]: Bit uneasy about the lack of a `tryCount` check in this termination condition, but I do suppose this has to converge in a finite number of iterations because of the constraints placed on `distance` (must be less than `distance_cap`, is geometrically shrunk every iteration).

Also I imagine this `checkForOverlap`'s overhead will be nasty (quadratic) due to how many objects are being skipped using linq. Something like a limited size queue would work better here.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:411]: `precedingIndex`

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:471]:
```suggestion
            if (almostEquals(nodeIndex, Math.Round(nodeIndex)))
```

**Review by bdach — DISMISSED** [review]
better readability-wise, but still needs work

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:246]: Yes, this is intentional. Hit objects should copy combo indices from an original HO at the same start time if possible, so `obj.StartTime == closestOrigObj.StartTime` is expected.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:294]: Not sure about this. Using `Interpolation.Lerp` would result in very similar code to the implementation of `map`, and `map` is a pretty standard name for a function that maps a value from one range to another imo.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:411]: A question: should I use `almostBigger` for `h.StartTime < time` here? I know that `h.StartTime` cannot be close to `time` but the reason for that lies in `getSamplesAtTime` and can take quite a while to comprehend.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: I have changed some calls to use `definitelyBigger`, mostly for readability's sake. I already considered the equality cases when using `almostBigger`, which was why some calls were strangely inverted.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (Henry-YSLin) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: The `SkipLast().TakeLast()` combo should run in constant time as long as the length of the resulting enumeration is constant (`overlap_check_count` in this case), so it shouldn't be a huge worry.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:294]: `map` is an overly generic name. You can map almost anything to anything. java/functional languages use that as a synonym to C# `.Select()`. `mapRange` would be better by me if you insist on using that.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs]: I'm not sure how `<` vs `almostBigger` matters here. So plain `<` for readability, probably.

**Comment by bdach** [comment]
I could be fine with this diff, but there is one more matter to resolve - `TestSceneModSelectOverlay.TestOsuMods` is failing because it was relying on the target mod being unimplemented. Unsure whether to just nuke the offending assert, or try to add coverage for unimplemented mods being unselectable in another manner.

**Comment by Henry-YSLin** [comment]
Creating a mock ruleset with a mock unimplemented mod is the safest way to test this imo. Not sure where to put these mock classes though.

**Comment by bdach** [comment]
Having those as inner classes of the test scene may be fine. It'd probably best be done in another separate pull, because this one is big enough on its own.

**Comment by Henry-YSLin** [comment]
@bdach A friendly ping: the test failure has been fixed. The current CI failure should be unrelated to this PR.

**Comment by bdach** [comment]
I'll dismiss my review, but this has been through so many rounds of review that I'm not super willing to merge without a double check from someone else.

**Comment by peppy** [comment]
I think overall the code quality seems fine for this, but I'm curious where you got the dim/fade behaviour from. This dim is something that needs to be implemented at a ruleset level, so I'm not sure I agree with it being done locally, but even if it is done here, the dim values seem completely random (grey 0.45).

**Comment by Henry-YSLin** [comment]
Not sure how this should be implemented at a ruleset level. Isn't the dim behaviour specific to this mod? 

As for the dim values, I got that from the cutting-edge implementation:
```csharp
Color col1 = new Color((byte)Math.Max(0, Colour.R * 0.45F), (byte)Math.Max(0, Colour.G * 0.45F), (byte)Math.Max(0, Colour.B * 0.45F), 255);
```

**Comment by peppy** [comment]
Right, thanks. I somehow missed that code. I guess this is actually applying some heavier dim than usual. The discrepancy is likely due to the way colour application is different in lazer and stable.

**Comment by peppy** [comment]
The fade in timings don't match stable's. It's important they do, as they are aligned with the beat (fade is exactly one beat earlier, so you have some visual cues). ie. `Preempt` should not be used at all.

**Comment by peppy** [comment]
The issue is you're using `PreemptTime` in the absolute sequence. You should be starting your sequence from `hitObject.StartTime - beatLength - 96` or something like that.

**Comment by Henry-YSLin** [comment]
But the entire sequence should be
- apply fade at `hitObject.StartTime - hitObject.TimePreempt`
- gradually remove fade from `hitObject.StartTime - beatLength - 96` to `hitObject.StartTime - beatLength`

Shouldn't it start at `hitObject.StartTime - hitObject.TimePreempt` then?

**Comment by peppy** [comment]
Sure, that sounds about right.

**Comment by Henry-YSLin** [comment]
> Shouldn't it start at `hitObject.StartTime - hitObject.TimePreempt` then?

That is the current behaviour. I was just wondering whether 96ms is accurate since it feels off in gameplay.

**Comment by peppy** [comment]
Okay, it's just written so weirdly it looks wrong. I'll refactor it.

**Comment by peppy** [comment]
One remaining issue I've noticed is the metronome plays twice near the first object, and doesn't play a countdown before the first object.

**Comment by Henry-YSLin** [comment]
The metronome playing twice at the beginning is caused by `BeatSyncedContainer` always firing `OnNewBeat` when the current time of the beatmap track is 0 and the beat index is negative. Is this intended behaviour? If not, it should probably be fixed in a separate PR.

As for the countdown, cutting-edge seems to extend the intro time to fit one bar of countdown for the metronome. No idea how to do that just yet.

**Comment by peppy** [comment]
I've fixed the remaining issues in this PR. Should be good to go.

My changes were minimal so I'm just going to get this one in. The other PR #13890 can come as a follow-up (I will apply the changes required to this mod in that PR.

**Review by peppy — DISMISSED** [review]

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (bdach) [inline osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:164]: Should this be out here like this, rather than inside the `startTime - preempt` absolute sequence? All other usages have such hide logic inside a transform sequence.

**Review by bdach — COMMENTED** [review]
Skimmed through @peppy's latest changes and all except maybe one look OK.

File: osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs:164]: I think due to what this is doing it's fine as-is, but let's adjust it for consistency.

**Review by bdach — APPROVED** [review]

**Pull request merged**
